File: Emulator/src/main/content/global/skill/construction/ConstructionInterface.java
package content.global.skill.construction;

import core.cache.def.impl.ItemDefinition;
import core.game.component.Component;
import core.game.component.ComponentDefinition;
import core.game.component.ComponentPlugin;
import core.game.node.entity.player.Player;
import core.game.node.entity.skill.Skills;
import core.game.node.scenery.Scenery;
import core.plugin.Initializable;
import core.plugin.Plugin;
import core.tools.Log;
import org.rs.consts.Components;

import static core.api.ContentAPIKt.log;

/**
 * The Construction interface.
 */
@Initializable
public final class ConstructionInterface extends ComponentPlugin {

	@Override
	public Plugin<Object> newInstance(Object arg) {
		ComponentDefinition.put(Components.POH_BUILD_FURNITURE_396, this);
		ComponentDefinition.put(Components.POH_HOUSE_OPTIONS_398, this);
		ComponentDefinition.put(Components.POH_BUILD_SCREEN_402, this);
		return this;
	}

	@Override
	public boolean handle(Player player, Component component, int opcode, int button, int slot, int itemId) {
		switch (component.id) {
			case Components.POH_BUILD_FURNITURE_396:
				switch (button) {
					case 132:
						player.getInterfaceManager().close();
						Hotspot hotspot = player.getAttribute("con:hotspot");
						Scenery object = player.getAttribute("con:hsobject");

						if (hotspot.getHotspot() != BuildHotspot.FLATPACK) {
							if (hotspot == null || object == null) {
								log(this.getClass(), Log.ERR, "Failed building decoration " + hotspot + " : " + object);
								break;
							}
						}

						slot = ((slot % 2 != 0) ? 4 : 0) + (slot >> 1);
						if (slot >= hotspot.getHotspot().getDecorations().length) {
							log(this.getClass(), Log.ERR,  "Failed building decoration " + slot + "/" + hotspot.getHotspot().getDecorations().length);
							break;
						}

						boolean debug = player.isStaff();
						Decoration deco = hotspot.getHotspot().getDecorations()[slot];

						if (hotspot.getHotspot() == BuildHotspot.FLATPACK) {
							deco = Decoration.forInterfaceItemId(itemId);
							player.sendMessage("Building flatpack: " + itemId);
							player.sendMessage("Building flatpack: " + deco.name());

							if(debug || checkRequirements(player, deco)) BuildingUtils.createFlatpack(player, deco, debug);
							break;
						}

						if (!debug) {
							if (player.getSkills().getLevel(Skills.CONSTRUCTION) < deco.getLevel()) {
								player.sendMessage("You need to have a Construction level of " + deco.getLevel() + " to build that.");
								return true;
							}
							if (!player.getInventory().containsItems(deco.getItems())) {
								player.sendMessage("You don't have the right materials.");
								return true;
							}
							for (int tool : deco.getTools()) {
								if (tool == BuildingUtils.WATERING_CAN) {
									boolean hasWateringCan = false;
									for (int i = 0; i < 8; i++) {
										if (player.getInventory().contains(tool - i, 1)) {
											hasWateringCan = true;
											break;
										}
									}
									if (!hasWateringCan) {
										player.sendMessage("You need a watering can to plant this.");
										return true;
									}
									continue;
								}
								if (!player.getInventory().contains(tool, 1)) {
									player.sendMessage("You need a " + ItemDefinition.forId(tool).getName() + " to build this.");
									return true;
								}
							}
						}
						BuildingUtils.buildDecoration(player, hotspot, deco, object);
						return true;
				}
				break;
			case Components.POH_HOUSE_OPTIONS_398:
				switch (button) {
					case 14:
						player.getHouseManager().toggleBuildingMode(player, true);
						return true;
					case 1:
						player.getHouseManager().toggleBuildingMode(player, false);
						return true;
					case 15:
						player.getHouseManager().expelGuests(player);
						return true;
					case 13:
						if (!player.getHouseManager().isInHouse(player)) {
							player.sendMessage("You can't do this outside of your house.");
							return true;
						}
						HouseManager.leave(player);
						return true;
				}
				break;
			case Components.POH_BUILD_SCREEN_402:
				int index = button - 160;
				log(this.getClass(), Log.FINE, "BuildRoom Interface Index: " + index);
				if (index > -1 && index < RoomProperties.values().length) {
					player.getDialogueInterpreter().open("con:room", RoomProperties.values()[index]);
					return true;
				}
				break;
		}
		return false;
	}

	private boolean checkRequirements(Player player, Decoration deco) {
		if (player.getSkills().getLevel(Skills.CONSTRUCTION) < deco.getLevel()) {
			player.sendMessage("You need to have a Construction level of " + deco.getLevel() + " to build that.");
			return true;
		}
		if (!player.getInventory().containsItems(deco.getItems())) {
			player.sendMessage("You don't have the right materials.");
			return true;
		}
		for (int tool : deco.getTools()) {
			if (tool == BuildingUtils.WATERING_CAN) {
				boolean hasWateringCan = false;
				for (int i = 0; i < 8; i++) {
					if (player.getInventory().contains(tool - i, 1)) {
						hasWateringCan = true;
						break;
					}
				}
				if (!hasWateringCan) {
					player.sendMessage("You need a watering can to plant this.");
					return false;
				}
				continue;
			}
			if (!player.getInventory().contains(tool, 1)) {
				player.sendMessage("You need a " + ItemDefinition.forId(tool).getName() + " to build this.");
				return false;
			}
		}
		return true;
	}
}

File: Server/src/main/content/global/skill/construction/ConstructionInterface.kt
package content.global.skill.construction

import core.cache.def.impl.ItemDefinition
import core.game.component.Component
import core.game.component.ComponentDefinition
import core.game.component.ComponentPlugin
import core.game.node.entity.player.Player
import core.game.node.entity.skill.Skills
import core.game.node.scenery.Scenery
import core.plugin.Initializable
import core.plugin.Plugin
import core.tools.Log
import shared.consts.Components
import core.api.log

@Initializable
class ConstructionInterface : ComponentPlugin() {

    override fun newInstance(arg: Any?): Plugin<Any> {
        ComponentDefinition.put(Components.POH_BUILD_FURNITURE_396, this)
        ComponentDefinition.put(Components.POH_HOUSE_OPTIONS_398, this)
        ComponentDefinition.put(Components.POH_BUILD_SCREEN_402, this)
        return this
    }

    override fun handle(player: Player, component: Component, opcode: Int, button: Int, slot: Int, itemId: Int): Boolean {
        when (component.id) {
            Components.POH_BUILD_FURNITURE_396 -> when (button) {
                132 -> {
                    player.interfaceManager.close()
                    val hotspot: Hotspot? = player.getAttribute("con:hotspot")
                    val obj: Scenery? = player.getAttribute("con:hsobject")

                    if (hotspot?.hotspot != BuildHotspot.FLATPACK) {
                        if (hotspot == null || obj == null) {
                            log(javaClass, Log.ERR, "Failed building decoration $hotspot : $obj")
                            return true
                        }
                    }

                    var slots = (if (slot % 2 != 0) 4 else 0) + (slot shr 1)
                    if (slots >= hotspot!!.hotspot.decorations.size) {
                        log(javaClass, Log.ERR, "Failed building decoration $slots/${hotspot.hotspot.decorations.size}")
                        return true
                    }

                    val debug = player.isStaff
                    var deco = hotspot.hotspot.decorations[slots]

                    if (hotspot.hotspot == BuildHotspot.FLATPACK) {
                        deco = Decoration.forInterfaceItemId(itemId)
                        if (debug || checkRequirements(player, deco)) {
                            BuildingUtils.createFlatpack(player, deco, debug)
                        }
                        return true
                    }

                    if (!debug) {
                        if (player.skills.getLevel(Skills.CONSTRUCTION) < deco.level) {
                            player.sendMessage("You need to have a Construction level of ${deco.level} to build that.")
                            return true
                        }
                        if (!player.inventory.containsItems(*deco.items)) {
                            player.sendMessage("You don't have the right materials.")
                            return true
                        }
                        for (tool in deco.tools) {
                            if (tool == BuildingUtils.WATERING_CAN) {
                                var hasWateringCan = false
                                for (i in 0..7) {
                                    if (player.inventory.contains(tool - i, 1)) {
                                        hasWateringCan = true
                                        break
                                    }
                                }
                                if (!hasWateringCan) {
                                    player.sendMessage("You need a watering can to plant this.")
                                    return true
                                }
                                continue
                            }
                            if (!player.inventory.contains(tool, 1)) {
                                player.sendMessage("You need a ${ItemDefinition.forId(tool).name} to build this.")
                                return true
                            }
                        }
                    }
                    BuildingUtils.buildDecoration(player, hotspot, deco, obj!!)
                    return true
                }
            }

            Components.POH_HOUSE_OPTIONS_398 -> when (button) {
                14 -> {
                    player.houseManager.toggleBuildingMode(player, true)
                    return true
                }
                1 -> {
                    player.houseManager.toggleBuildingMode(player, false)
                    return true
                }
                15 -> {
                    player.houseManager.expelGuests(player)
                    return true
                }
                13 -> {
                    if (!player.houseManager.isInHouse(player)) {
                        player.sendMessage("You can't do this outside of your house.")
                        return true
                    }
                    HouseManager.leave(player)
                    return true
                }
            }

            Components.POH_BUILD_SCREEN_402 -> {
                val index = button - 160
                log(javaClass, Log.FINE, "BuildRoom Interface Index: $index")
                if (index in RoomProperties.values().indices) {
                    player.dialogueInterpreter.open("con:room", RoomProperties.values()[index])
                    return true
                }
            }
        }
        return false
    }

    private fun checkRequirements(player: Player, deco: Decoration): Boolean {
        if (player.skills.getLevel(Skills.CONSTRUCTION) < deco.level) {
            player.sendMessage("You need to have a Construction level of ${deco.level} to build that.")
            return false
        }
        if (!player.inventory.containsItems(*deco.items)) {
            player.sendMessage("You don't have the right materials.")
            return false
        }
        for (tool in deco.tools) {
            if (tool == BuildingUtils.WATERING_CAN) {
                var hasWateringCan = false
                for (i in 0..7) {
                    if (player.inventory.contains(tool - i, 1)) {
                        hasWateringCan = true
                        break
                    }
                }
                if (!hasWateringCan) {
                    player.sendMessage("You need a watering can to plant this.")
                    return false
                }
                continue
            }
            if (!player.inventory.contains(tool, 1)) {
                player.sendMessage("You need a ${ItemDefinition.forId(tool).name} to build this.")
                return false
            }
        }
        return true
    }
}
